src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,291 @@
+package consistenthashing
+
+import (
+	"context"
+	"fmt"
+	"hash/crc32"
+	"path"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/dlock"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+
+	"github.com/sirupsen/logrus"
+
+	etcd "go.etcd.io/etcd/client/v3"
+)
+
+// Ring is a consistent hash Ring. Each process should only create one instance with a given prefix. Nodes can be
+// added locally to a Ring. A Ring watches for changes to the prefix to determine if new nodes have been added
+// or deleted.
+// A Ring should be initialized with a call to New.
+type Ring struct {
+	client    *etcd.Client
+	hashFn    func([]byte) uint32
+	stateLock sync.Mutex
+	logger    *logrus.Logger
+	members   []Member
+	node      node   // node represents the member instance in a Ring that is local to the current process.
+	prefix    string // prefix is prepended to node keys stored in etcd to avoid naming conflicts.
+}
+
+// node is a local Member of a Ring. Upon creation, a node requests a lease from etcd and continuously refreshes it
+// until it is shutdown.
+type node struct {
+	Member
+	Locks map[string]lockInfo
+}
+
+// Member contains metadata about members in a Ring.
+type Member struct {
+	Id   string
+	Hash uint32
+}
+
+// lockInfo instances are created for keys by a ring's node.
+// A context lockInfo.Ctx is stored so the proper context is used when unlocking lockInfo.Lock
+type lockInfo struct {
+	Lock dlock.DLock
+	Ctx  context.Context
+}
+
+// New is the default constructor for Ring. Creating a Ring starts a goroutine to watch prefix in etcd and creates a
+// node that continuously refreshes a lease in etcd.
+func New(ctx context.Context, client *etcd.Client, logger *logrus.Logger, prefix string, id ...string) (*Ring, error) {
+	if id == nil {
+		id = []string{uuid.New()}
+	}
+	localMember := Member{
+		Id:   id[0],
+		Hash: crc32.ChecksumIEEE([]byte(id[0])),
+	}
+	ring := &Ring{
+		client:  client,
+		hashFn:  crc32.ChecksumIEEE,
+		logger:  logger,
+		members: []Member{localMember},
+		node: node{
+			Member: localMember,
+			Locks:  map[string]lockInfo{},
+		},
+		prefix:    prefix,
+		stateLock: sync.Mutex{},
+	}
+	etcdCol := collection.NewEtcdCollection(ring.client, path.Join(ring.prefix, "nodes"),
```

Comment:
We can just use empty type rather than bool, also the arbitrary newline is a bit weird. I am generally in favor of letting the lines run long to put pressure on reducing the number of parameters to a function when that occurs.